src/constraint.rs

```diff
@@ -1406,3 +1420,67 @@ impl ConstraintValidator for ScaleConstraint {
         Ok(())
     }
 }
+
+/// Implements Ion Schema's `timestamp_precision` constraint
+/// [scale]: https://amzn.github.io/ion-schema/docs/spec.html#timestamp_precision
+#[derive(Debug, Clone, PartialEq)]
+pub struct TimestampPrecisionConstraint {
+    timestamp_precision_range: Range,
+}
+
+impl TimestampPrecisionConstraint {
+    pub fn new(scale_range: Range) -> Self {
+        Self {
+            timestamp_precision_range: scale_range,
+        }
+    }
+
+    pub fn timestamp_precision(&self) -> &Range {
+        &self.timestamp_precision_range
+    }
+}
+
+impl ConstraintValidator for TimestampPrecisionConstraint {
+    fn validate(&self, value: &OwnedElement, type_store: &TypeStore) -> ValidationResult {
+        // get the precision of given timestamp
+        let timestamp_value = match value.as_timestamp() {
+            Some(timestamp_value) => timestamp_value,
+            _ => {
+                // return Violation if value is not timestamp
+                let error_message = if value.is_null() {
+                    format!("expected a timestamp but found {:?}", value)
+                } else {
+                    format!("expected a timestamp but found {}", value.ion_type())
+                };
+                return Err(Violation::new(
+                    "timestamp_precision",
+                    ViolationCode::TypeMismatched,
+                    &error_message,
+                ));
+            }
+        };
+
+        let value_precision = TimestampPrecisionValue::from_timestamp(&timestamp_value);
```

Comment:
issue—Clippy says "this expression creates a reference which is immediately dereferenced by the compiler".